pxf/pxf-json/src/main/java/org/apache/hawq/pxf/plugins/json/JsonInputFormat.java

```diff
@@ -0,0 +1,226 @@
+		}
+
+		@Override
+		public boolean next(Text key, NullWritable value) throws IOException {
+			if (rdr.next(this.key, this.value)) {
+				key.set(this.value);
+				return true;
+			} else {
+				return false;
+			}
+		}
+
+		@Override
+		public float getProgress() throws IOException {
+			return rdr.getProgress();
+		}
+	}
+
+	public static class JsonRecordReader implements RecordReader<Text, NullWritable> {
+
+		private Logger LOG = Logger.getLogger(JsonRecordReader.class);
+
+		private JsonStreamReader rdr = null;
+		private long start = 0, end = 0;
+		private float toRead = 0;
+		private String identifier = null;
+		private Logger log = Logger.getLogger(JsonRecordReader.class);
+
+		public JsonRecordReader(JobConf conf, FileSplit split) throws IOException {
+			log.info("JsonRecordReader constructor called.  Conf is " + conf + ". Split is " + split);
+			this.identifier = conf.get(RECORD_IDENTIFIER);
+			log.info("Identifier is " + this.identifier);
+
+			if (this.identifier == null || identifier.isEmpty()) {
+				throw new InvalidParameterException(JsonInputFormat.RECORD_IDENTIFIER + " is not set.");
+			} else {
+				LOG.info("Initializing JsonRecordReader with identifier " + identifier);
+			}
+
+			// get relevant data
+			Path file = split.getPath();
+
+			log.info("File is " + file);
+
+			start = split.getStart();
+			end = start + split.getLength();
+			toRead = end - start;
+			log.info("FileSystem is " + FileSystem.get(conf));
+
+			FSDataInputStream strm = FileSystem.get(conf).open(file);
+
+			log.info("Retrieved file stream ");
+
+			if (start != 0) {
+				strm.seek(start);
+			}
+
+			rdr = new JsonStreamReader(identifier, new BufferedInputStream(strm));
+
+			log.info("Reader is " + rdr);
+		}
+
+		@Override
+		public boolean next(Text key, NullWritable value) throws IOException {
+
+			boolean retval = false;
+			boolean keepGoing = false;
+			do {
+				// Exit condition (end of block/file)
+				if (rdr.getBytesRead() >= (end - start)) {
+					return false;
+				}
+
+				keepGoing = false;
+				String record = rdr.getJsonRecord();
+				if (record != null) {
+					if (JsonInputFormat.decodeLineToJsonNode(record) == null) {
+						log.error("Unable to parse JSON string.  Skipping. DEBUG to see");
+						log.debug(record);
+						keepGoing = true;
+					} else {
+						key.set(record);
+						retval = true;
+					}
+				}
+			} while (keepGoing);
+
+			return retval;
+		}
+
+		@Override
+		public Text createKey() {
+			return new Text();
+		}
+
+		@Override
+		public NullWritable createValue() {
+			return NullWritable.get();
+		}
+
+		@Override
+		public long getPos() throws IOException {
+			return start + rdr.getBytesRead();
+		}
+
+		@Override
+		public void close() throws IOException {
+			rdr.close();
+		}
+
+		@Override
+		public float getProgress() throws IOException {
+			return (float) rdr.getBytesRead() / toRead;
+		}
+	}
+
+	public static synchronized JsonNode decodeLineToJsonNode(String line) {
```

Comment:
why does the mapper have to be static? wouldn't it be better to make it a member of the JsonInputFormat, and then not worry about synchronizing this function?
When handling multiple requests, it probably has a performance implication.